Proofs/Experiments/MemoryAliasing.lean

```diff
@@ -33,14 +121,206 @@ theorem mem_automation_test_2
                   dest_addr (n0 <<< 4)) :
   read_mem_bytes 16 src_addr (write_mem_bytes 16 dest_addr blah s0) =
   read_mem_bytes 16 src_addr s0 := by
-  sorry
-  -- rw [read_mem_bytes_of_write_mem_bytes_different (by decide) (by decide)]
-  -- rwa [@mem_separate_for_subset_general
-  --           src_addr (src_addr + (n0 <<< 4 - 1))
-  --           dest_addr (dest_addr + (n0 <<< 4 - 1))
-  --           src_addr (src_addr + 15#64)
-  --           dest_addr (dest_addr + 15#64)]
-  -- repeat sorry
-
-/-- info: 'mem_automation_test_2' depends on axioms: [propext, sorryAx, Quot.sound] -/
+  simp only [memory_rules]
+  simp_mem
+  rfl
+
+
+/-- info: 'mem_automation_test_2' depends on axioms: [propext, Classical.choice, Quot.sound] -/
 #guard_msgs in #print axioms mem_automation_test_2
+
+
+namespace ReadOverlappingRead
+
+/-- A read overlapping with another read. -/
+theorem overlapping_read_test_1 {out : BitVec (16 * 8)}
+    (hlegal : mem_legal' src_addr 16)
+    (h : read_mem_bytes 16 src_addr s = out) :
+    read_mem_bytes 16 src_addr s = out := by
+  simp only [memory_rules] at h ⊢
+  simp_mem
+  simp only [Nat.reduceMul, Nat.sub_self, BitVec.extractLsBytes_eq_self]
+
+/--
+info: 'ReadOverlappingRead.overlapping_read_test_1' depends on axioms: [propext, Classical.choice, Quot.sound]
+-/
+#guard_msgs in #print axioms overlapping_read_test_1
+
+/-- A read overlapping with another read. -/
+theorem overlapping_read_test_2 {out : BitVec (16 * 8)}
+    (hlegal : mem_legal' src_addr 16)
+    (h : read_mem_bytes 16 src_addr s = out) :
+    read_mem_bytes 10 (src_addr + 6) s = out.extractLsBytes 6 10 := by
+  simp only [memory_rules] at h ⊢
+  simp_mem
+  · congr
+    -- ⊢ (src_addr + 6).toNat - src_addr.toNat = 6
+    bv_omega'
+/--
+info: 'ReadOverlappingRead.overlapping_read_test_2' depends on axioms: [propext, Classical.choice, Quot.sound]
+-/
+#guard_msgs in #print axioms overlapping_read_test_2
+
+/-- A read in the goal state overlaps with a read in the
+left hand side of the hypotheis `h`.
+-/
+theorem overlapping_read_test_3
+    (hlegal : mem_legal' src_addr 16)
+    (h : read_mem_bytes 16 src_addr s = read_mem_bytes 16 other_addr s) :
+    read_mem_bytes 10 (src_addr + 6) s =
+    -- @bollu: unfortunate, this needs `s.mem` to be public. Annoying.
+      (Memory.read_bytes 16 other_addr s.mem).extractLsBytes 6 10 := by
+  simp only [memory_rules] at h ⊢
+  simp_mem
+  · congr
+    -- ⊢ (src_addr + 6).toNat - src_addr.toNat = 6
+    bv_omega'
+/--
+info: 'ReadOverlappingRead.overlapping_read_test_3' depends on axioms: [propext, Classical.choice, Quot.sound]
+-/
+#guard_msgs in #print axioms overlapping_read_test_3
+
+/- TODO(@bollu): This test case hangs at `bv_omega'`. This is to be debugged.
```

Comment:
Do we need to make issues to track these TODOs?

Reply:
https://github.com/leanprover/LNSym/issues/116